src/socket/icmp.rs

```diff
@@ -0,0 +1,569 @@
+        } else {
+            Err(Error::Truncated)
+        }
+    }
+}
+
+impl<'a> Resettable for PacketBuffer<'a> {
+    fn reset(&mut self) {
+        self.size = 0;
+    }
+}
+
+/// An ICMPv4 packet ring buffer.
+pub type SocketBuffer<'a, 'b: 'a> = RingBuffer<'a, PacketBuffer<'b>>;
+
+/// An ICMPv4 socket
+///
+/// An ICMPv4 socket is bound to a specific [IcmpEndpoint] which may
+/// be a sepecific UDP port to listen for ICMP error messages related
+/// to the port or a specific ICMP identifier value. See [bind] for
+/// more details.
+///
+/// [IcmpEndpoint]: enum.IcmpEndpoint.html
+/// [bind]: #method.bind
+#[derive(Debug)]
+pub struct IcmpSocket<'a, 'b: 'a> {
+    handle:    SocketHandle,
+    rx_buffer: SocketBuffer<'a, 'b>,
+    tx_buffer: SocketBuffer<'a, 'b>,
+    /// The endpoint this socket is communicating with
+    endpoint:  Endpoint,
+    /// The time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
+    ttl:       Option<u8>
+}
+
+impl<'a, 'b> IcmpSocket<'a, 'b> {
+    /// Create an ICMPv4 socket with the given buffers.
+    pub fn new(rx_buffer: SocketBuffer<'a, 'b>, tx_buffer: SocketBuffer<'a, 'b>) -> Socket<'a, 'b> {
+        Socket::Icmp(IcmpSocket {
+            handle:    SocketHandle::EMPTY,
+            rx_buffer: rx_buffer,
+            tx_buffer: tx_buffer,
+            endpoint:  Endpoint::default(),
+            ttl:       None
+        })
+    }
+
+    /// Return the socket handle.
+    #[inline]
+    pub fn handle(&self) -> SocketHandle {
+        self.handle
+    }
+
+    /// Set the socket handle.
+    pub(in super) fn set_handle(&mut self, handle: SocketHandle) {
+        self.handle = handle;
+    }
+
+    /// Return the time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
+    ///
+    /// See also the [set_ttl](#method.set_ttl) method
+    pub fn ttl(&self) -> Option<u8> {
+        self.ttl
+    }
+
+    /// Set the time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
+    ///
+    /// A socket without an explicitly set TTL value uses the default [IANA recommended]
+    /// value (64).
+    ///
+    /// # Panics
+    ///
+    /// This function panics if a TTL value of 0 is given. See [RFC 1122 § 3.2.1.7].
+    ///
+    /// [IANA recommended]: https://www.iana.org/assignments/ip-parameters/ip-parameters.xhtml
+    /// [RFC 1122 § 3.2.1.7]: https://tools.ietf.org/html/rfc1122#section-3.2.1.7
+    pub fn set_ttl(&mut self, ttl: Option<u8>) {
+        // A host MUST NOT send a datagram with a Time-to-Live (TTL) value of 0
+        if let Some(0) = ttl {
+            panic!("the time-to-live value of a packet must not be zero")
+        }
+
+        self.ttl = ttl
+    }
+
+    /// Bind the socket to the given endpoint.
+    ///
+    /// This function returns `Err(Error::Illegal)` if the socket was open
+    /// (see [is_open](#method.is_open)), and `Err(Error::Unaddressable)`
+    /// if `endpoint` is `IpEndpoint` and the port is zero or if `endpoint`
```

Comment:
I'd say something like:

    if `endpoint` is unspecified (see [is_unspecified])